[DOCS-6433] Add filter logs with subqueries (#21102)
DataDog/documentation: content/en/logs/explorer/advanced_filtering.md

Using `service:network_directory` as your main query, you can then filter out results from recognized devices by defining a subquery filter for your `device-manager` service like this:

{{< img src="logs/explorer/advanced_filtering/narrow_dataset_example.png" alt="" style="width:100%;" >}}

## Filter logs based on Reference Tables

<div class="alert alert-warning">Reference Tables and the feature to filter logs based on Reference Tables are in beta. There is a limit of 100 Reference Tables per account and a limit of 1000 rows in a Reference Table.</div>

Reference Tables allow you to combine metadata with logs, providing more information to resolve application issues. Add a query filter based on a Reference Table to perform lookup queries. For more information on creating and managing this feature, see the [Reference Tables][2] guide.

To apply a query filter with Reference Tables, click on the `+ Add` button next to the query editor and select **Reference Table**. In the following example, the Reference Table query filter is used to search all recent logs that include a malicious IP address from a threat intel reference table:

> {{< img src="/logs/explorer/group/reference_tables_logs.png" alt="The Datadog log explorer with reference table search options highlighted" border="true" popup="true" style="width:100%;" >}}

Author: I'll move the image to the new folder in a separate PR.

1. Select whether you want to use the **IN** or **NOT IN** operator:  
    - The results from the **IN** operator only include logs where the attribute value is also found in the subquery results. For example, if you only want to see logs generated by `service:a` and is associated to a user who is also one of the top users of `service:b`.  
    - The results from the **NOT IN** operator exclude logs where the attribute value is found in the subquery results. For example, if you only want to see `status:error` logs, but also want to filter out `status:error` logs where users associated with those logs eventually end up with a `status:success` log. See [Filter outdated or superseded logs](#filter-outdated-or-superseded-logs) for a detailed example.  
1. Optionally, reduce the number of subquery attribute values to match against. The default value and maximum value is `1000`. Choose between **top** (highest frequency values) or **bottom** (lowest frequency values).

### Subquery examples

The following are scenarios where you need to use subqueries to get the information you need from your logs.

> #### Filter outdated or superseded logs

Reviewer: ```suggestion
#### Filter outdated or superseded logs
```
For some reason I don't see this sub-heading on the TOC under the H3 "Subquery examples" ?

Author: Hm, I checked other docs and it looks like TOCs only list up to H3, so it's expected?

Assume you operate an e-commerce platform. A log is generated each time one of your customers tries to place an order. You want to analyze your logs to understand the total value of potential purchases lost due to an ongoing issue with your website.

You realize, however, that an order can fail multiple times before successfully completing: meaning for that particular order ID, there are log entries in the search results for both `status:error` logs and `status:success` logs. If you extracted lists of unique order IDs from the two queries, this order ID would appear in both. With subqueries, you can obtain a mutually exclusive list.

> In this example, you are only interested in logs for orders that did not eventually succeed. To filter out orders that eventually succeeded using the subquery feature:

Reviewer: ```suggestion
In this example, you are only interested in logs for orders that did not eventually succeed. To exclude orders that eventually succeeded using the subquery feature:
```
I had to re-read this a few times to understand the meaning, so I think to the reader, the word "exclude" would make more sense vs "filter out", but kind of your preference :)

1. Define a subquery for `status:success` logs.
1. Select the **NOT IN** operator to exclude orders from the subquery result set.

{{< img src="logs/explorer/advanced_filtering/filter_outdated_example.png" alt="The query editor showing the setup for filtering out orders that ended up succeeding" style="width:100%;" >}}

> #### Correlate across different log sources

Reviewer: ```suggestion
#### Correlate across different log sources
```
This isn't showing on the TOC either FYI

Assume you have a service named `network_directory` that monitors all internal network resources and access to those resources within your organization. Log events generated by this service include standard attributes (like `host`, `service`, `source`) and custom attributes like the client's IP address.

> You have another `device-manager` service that tracks all internal assets (infrastructure, employee devices, and so on.)

Reviewer: ```suggestion
Additionally, you have another `device-manager` service that tracks all internal assets (infrastructure, employee devices, and so on.)
```

> You are investigating an ongoing attack and there is a significant increase in API requests across almost all of your endpoints. You want to first identify IP addresses associated with anomalous request volumes so that you can quickly block them at the firewall level. However, your internal services are some of the largest consumers of these endpoints and you need to exclude them from your query results to avoid mistakenly blocking them.

Reviewer: ```suggestion
You are investigating an ongoing attack and observe there is a significant increase in API requests across almost all of your endpoints. You want to first identify IP addresses associated with anomalous request volumes so that you can block them at the firewall level. However, your internal services are some of the largest consumers of these endpoints and you need to exclude them from your query results to avoid mistakenly blocking them.
```

> To do that, use `service:network_directory` as your main query, and then define a subquery filter for your `device-manager` service to filter out results from recognized devices.

Reviewer: ```suggestion
In this example, use `service:network_directory` as your main query, and then define a subquery filter for your `device-manager` service to filter out results from recognized devices.
```
